### ISSUE TITLE
Batch norm performance

### ISSUE THREAD

**Issue by BobMcDear** [BobMcDear/attorch#7]
Hi, thanks for your amazing work, but in my environment(Tesla V100), the kernel, `batch_norm_forward_kernel`, seems to be much slower than PyTorch's interface (avg 25x, autotune time has been removed) . Ex for input shape like: B, C, N=(32, 128, 16384). I just copy the test function of batch_norm and revise the shape parameter, did I do something wrong?

_Originally posted by @0Addicted0 in https://github.com/triton-lang/triton/issues/900#issuecomment-2500813279_

**Comment by BobMcDear** [comment]
Hello,

Thank you for bringing this up. Are you testing this under a mixed-precision setting? Triton kernels can be quite slow when using the default single-precision format, which is not common nowadays anyhow since most practitioners use mixed precision. Additionally, the performance of PyTorch and attorch modules varies wildly depending on the input shape. For example, attorch can be slower when fed low-dimensional, high-resolution data that appears in the early layers of a CNN, whereas it'll be faster with high-dimensional, low-resolution data typical of the later layers in the network. To get a better idea of attorch's true performance, I suggest running an end-to-end benchmark that accurately reflects the usage of batch norm in practice. The [Imagenette example](https://github.com/BobMcDear/attorch/tree/main/examples/imagenette) with ResNet would be an appropriate starting point.

**Comment by 0Addicted0** [comment]
``` bash
python -m examples.imagenette.main
``` 
Thank you for your quick reply. I ran the above command to test on ResNet101, and the results are as follows: it appears that the average forward and backward times are indeed better than those in PyTorch, but the total training time seems to be longer overall?

``` bash
attorch run:
Forward pass mean execution time: 14.736896514892578
Backward pass mean execution time: 23.92473602294922
Forward plus backward pass mean execution time: 38.6616325378418
Epoch 1/10
Training loss: 2.2186184620453138
Validation accuracy: 0.24154713114754098
Epoch 2/10
Training loss: 1.998960185859163
Validation accuracy: 0.33734631147540983
Epoch 3/10
Training loss: 1.8621821060019024
Validation accuracy: 0.4175204918032787
Epoch 4/10
Training loss: 1.7291115037465499
Validation accuracy: 0.44748975409836067
Epoch 5/10
Training loss: 1.6801815253193095
Validation accuracy: 0.4339139344262295
Epoch 6/10
Training loss: 1.6484606791350802
Validation accuracy: 0.5463627049180327
Epoch 7/10
Training loss: 1.515866575200679
Validation accuracy: 0.4884733606557377
Epoch 8/10
Training loss: 1.493094529539852
Validation accuracy: 0.5489241803278688
Epoch 9/10
Training loss: 1.4513899649603892
Validation accuracy: 0.5981045081967213
Epoch 10/10
Training loss: 1.5113259313470226
Validation accuracy: 0.5783811475409836
Total training and validation time: 279.1156380176544
PyTorch run:
Forward pass mean execution time: 14.308521270751953
Backward pass mean execution time: 26.591232299804688
Forward plus backward pass mean execution time: 40.89975357055664
Epoch 1/10
Training loss: 2.24118921958794
Validation accuracy: 0.24487704918032788
Epoch 2/10
Training loss: 2.008115643970037
Validation accuracy: 0.3201844262295082
Epoch 3/10
Training loss: 1.8938709453000862
Validation accuracy: 0.3824282786885246
Epoch 4/10
Training loss: 1.7935907056776144
Validation accuracy: 0.43621926229508196
Epoch 5/10
Training loss: 1.7025692875102414
Validation accuracy: 0.46234631147540983
Epoch 6/10
Training loss: 1.5878330246876862
Validation accuracy: 0.5586577868852459
Epoch 7/10
Training loss: 1.529219527163748
Validation accuracy: 0.5007684426229508
Epoch 8/10
Training loss: 1.4856143498824814
Validation accuracy: 0.5650614754098361
Epoch 9/10
Training loss: 1.4533872907444583
Validation accuracy: 0.6037397540983607
Epoch 10/10
Training loss: 1.4042550361762614
Validation accuracy: 0.5809426229508197
Total training and validation time: 244.4431610107422
```

**Comment by BobMcDear** [comment]
I can reproduce your results. The default batch size of 32 is quite small for the V100's memory and severely underutilizes GPU resources. If you set it to the largest value possible (128 in the case of ResNet-101), you should notice that attorch surpasses or exceeds the speed of PyTorch.

**Comment by 0Addicted0** [comment]
Thanks, I reproduced it, increase the`batch size` is usefull
